pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +234,295 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+	obj, quit := c.gatewayQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.gatewayQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+
+	if c.wireGuardConfig != nil {
+		isGatewayNode, err := c.reconcileWireGuard()
+		if err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+			return true
+		}
+
+		// Controller will try to list all existing ClusterInfoImports and add them as
+		// WireGuard peers when a WireGuard interface is initialized and the Node is
+		// Gateway Node.
+		if isGatewayNode {
+			if err := c.updateWireGuardPeers(""); err != nil {
+				klog.ErrorS(err, "Faield to update WireGuard peers")
+				c.gatewayQueue.AddRateLimited(gateway)
+				return true
+			}
+		}
+		klog.InfoS("Reconcile WireGuard successfully")
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
 	}
 	return true
 }
 
-func (c *MCRouteController) syncMCFlows() error {
+// reconcileWireGuard reconciles the WireGuard configurations. If Multi-cluster WireGuard feature enabled, it will do:
+//  1. Initialize corresponding WireGuard interface and IP route on the host, and add existing ClusterInfoImport to
+//     WireGuard peers when a new Gateway created.
+//  2. Remove WireGuard interface and IP route on the host when a Gateway deleted.
+//
+// reconcileWireGuard supports reentrance. If the initialization process failed partly, it will remove stale resources.
+func (c *MCDefaultRouteController) reconcileWireGuard() (bool, error) {
+	gateway, err := c.getActiveGateway()
+	if err != nil {
+		return false, err
+	}
+	c.wireGuardMutex.Lock()
+	defer c.wireGuardMutex.Unlock()
+	if c.wireGuardInitialized {
+		// The Gateway is deleted. Remove WireGuard interface
+		if gateway == nil || gateway.Name != c.nodeConfig.Name {
+			klog.InfoS("Gateway is deleted, removing WireGuard interface and rule")
+			if err := c.deleteWireGuardInterfaceAndRule(); err != nil {
+				klog.ErrorS(err, "Failed to delete interface and route")
+				return false, err
+			}
+			c.wireGuardInitialized = false
+			return false, err
+		}
+		return true, nil
+	}
+
+	// If the WireGuard is not initialized, and the Node is the Gateway Node.
+	// Controller will create corresponding interface and route on the host.
+	if !c.wireGuardInitialized && gateway != nil && gateway.Name == c.nodeConfig.Name {
```

Comment:
Consider define a boolean like: `isGateway := gateway != nil && gateway.Name == c.nodeConfig.Name`, so the code simpler, and we can use `!isGateway` in the previous `if c.wireGuardInitialized` block.